modules/pwa/class.jetpack-pwa-manifest.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+class Jetpack_PWA_Manifest {
+	/**
+	 * @var Jetpack_PWA_Manifest
+	 */
+	private static $__instance = null;
+
+	/**
+	 * When this query var is present, display the PWA manifest.
+	 *
+	 * @var string
+	 */
+	const PWA_MANIFEST_QUERY_VAR = 'jetpack_app_manifest';
+
+	/**
+	 * Singleton implementation
+	 *
+	 * @return Jetpack_PWA_Manifest
+	 */
+	public static function instance() {
+		if ( is_null( self::$__instance ) ) {
+			self::$__instance = new Jetpack_PWA_Manifest;
+		}
+
+		return self::$__instance;
+	}
+
+	/**
+	 * Registers actions the first time that instance() is called.
+	 */
+	private function __construct() {
+		add_action( 'wp_head', array( $this, 'render_manifest_link' ) );
+		add_action( 'amp_post_template_head', array( $this, 'render_manifest_link' ) );
+		add_action( 'template_redirect', array( $this, 'render_manifest_json' ), 2 );
+	}
+
+	function render_manifest_link() {
+		?>
+			<link rel="manifest" href="<?php echo esc_url_raw( $this->get_manifest_url() ); ?>">
+			<meta name="theme-color" content="<?php echo esc_attr( Jetpack_PWA_Helpers::get_theme_color() ); ?>">
+		<?php
+	}
+
+	public function get_manifest_url() {
+		return add_query_arg(
+			self::PWA_MANIFEST_QUERY_VAR, '1', home_url()
+		);
+	}
+
+	function render_manifest_json() {
+		// Do not load manifest in multiple locations
+		if ( is_front_page() && isset( $_GET[ self::PWA_MANIFEST_QUERY_VAR ] ) && $_GET[ self::PWA_MANIFEST_QUERY_VAR ] ) {
```

Comment:
`is_front_page()` is quite an expensive check, relatively speaking. What is the risk of just allowing this to fire if the `PWA_MANIFEST_QUERY_VAR` is present in the URL? Are we afraid that users might accidentally suffix it to other URLs?

Reply:
I don't think that `is_front_page()` is an expensive check. It does check several options, but it appears that they're autoloaded based on `wp-admin/schema.php`, so they'd be in memory.

The reasoning for only showing manifest from just the front page was so that we had one canonical URL to point to, as opposed to showing it on any page where the query parameter was present.